### ISSUE TITLE
Snowflake "Unknown function ENCODE" when casting to STRING

### ISSUE THREAD

**Issue by henrihem** [GoogleCloudPlatform/professional-services-data-validator#1070]
### Issue 1:
When running comparisong from Snowflake to BigQuery with the following configuration file:
```
result_handler: {}
source: sf_conn
target: bq_conn
validations:
- calculated_fields: []
  comparison_fields:
  - cast: string
    field_alias: dv_id
    source_column: DV_ID
    target_column: dv_id
  filter_status: null
  filters: []
  format: table
  labels: []
  primary_keys:
  - cast: null
    field_alias: key_field
    source_column: KEY_FIELD
    target_column: key_field
  random_row_batch_size: '100'
  schema_name: dv
  table_name: TABLE1
  target_schema_name: dv
  target_table_name: TABLE1
  threshold: 0.0
  type: Row
  use_random_rows: true
```

With casting dv_id as string:
`  comparison_fields:
  - cast: string
    field_alias: dv_id
    source_column: DV_ID
    target_column: dv_id`  

It will result in error:
`Unknown function ENCODE`

And the resulted SQL will be the following:
`SELECT encode(t1."DV_ID", %(encode_1)s)`

Correct format should be:
`SELECT hex_encode(t1."DV_ID", %(encode_1)s)`

Resulting in pure SQL:
`SELECT hex_encode(t1."DV_ID", 0)`

Now the pure SQL result is incorrectly this:
`SELECT encode(t1."DV_ID", 'escape')`

### Issue 2
Without casting it does work ok:
```
  comparison_fields:
  - cast: null
    field_alias: dv_id
    source_column: DV_ID
    target_column: dv_id
```
But if there is result_handler specified, it will fail in another error:
`PM-ERROR: Error Object of type bytes is not JSON serializable occurred while running config file test.yaml. Skipping it for now.
`

### Issue 3
Another one when comparing fields with Snowflake BINARY datatype and BigQuery BYTES. Queries go just fine to databases, but when writing results to result_handler, it will fail in error.

Configuration file:
```
result_handler:
  project_id: project_id
  table_id: table_id
  type: BigQuery
source: sf_conn
target: bq_conn
validations:
- calculated_fields: []
  comparison_fields:
  - cast: null
    field_alias: comp_field_1
    source_column: COMP_FIELD_1
    target_column: comp_field_1
  - cast: null
    field_alias: comp_field_1
    source_column: COMP_FIELD_1
    target_column: comp_field_1
  filter_status: null
  filters: []
  format: table
  labels: []
  primary_keys:
  - cast: null
    field_alias: dv_id
    source_column: DV_ID
    target_column: dv_id
  random_row_batch_size: '100'
  schema_name: dv
  table_name: TABLE1
  target_schema_name: dv
  target_table_name: TABLE1
  threshold: 0.0
  type: Row
  use_random_rows: true
```

Resulting error:
`12/12/2023 02:29:12 PM-ERROR: Error cannot use a string pattern on a bytes-like object occurred while running config file test2.yaml. Skipping it for now.`

**Comment by helensilva14** [comment]
Hi @henrihem! I'll start to investigate your issue, something that would be really helpful: could you please share the DDL statements for the Snowflake and BQ tables? With that I can better simulate your scenario, thanks!

**Comment by henrihem** [comment]
Hi @helensilva14 , sorry for late response.
Here is the DDL for the tables:
Snowflake:
`create or replace table dv.TABLE1 (
	DV_ID BINARY(16),
	KEY_FIELD VARCHAR(128)
);`

BigQuery
`CREATE TABLE dv.TABLE1 (
     dv_id BYTES(16),  
	 KEY_FIELD STRING(128)
);`

**Comment by henrihem** [comment]
Also if I run this one:
```
data-validation validate row \
  -sc bq_conn \
  -tc sf_conn \
  -tbls dv.TABLE1=dv.TABLE1 \
  --concat key_field \
  --primary-keys dv_id \
  -rr \
  -rbs 100
```
  
  It fails in:
  `TypeError: cannot use a string pattern on a bytes-like object`

**Comment by sundar-mudupalli-work** [comment]
Henri,

I have been able to reproduce your problem(s). Can you outline exactly what you are trying to achieve ?

I have been able to successfully perform a `validate row -comp-fields DV_ID -pk KEY_FIELD -rr -rbs 10` with your table schema between Snowflake and BigQuery (with KEY_FIELD as primary key). I am not able to store that vaildation result in BigQuery due to the `TypeError: Object of type bytes is not JSON serializable` - looks like we are trying to serialize the DV_ID field - so need to check how to fix this issue.

Our objective is not to fix DVT issues. Our objective is to get customers successful in their efforts to use DVT to validate two tables. The BigQuery issue is likely a bug - if fixed will help you. The other challenges you report (string in the yaml file causing ENCODE etc), I don't understand how they are related to the problem you are trying to solve - because I don't know exactly what your use case is.

If you can share that it will be helpful for us to prioritize and find fixes for you.

Thanks.

Sundar Mudupalli

**Comment by helensilva14** [comment]
Hi everyone! I opened a draft PR https://github.com/GoogleCloudPlatform/professional-services-data-validator/pull/1113 with our initial approach to solve this issue and that's the situation I'm at:

**Case 1**
I couldn't find yet which method can produce the desired output. Only the ENCODE functions worked so far but values are not matching though.

- BASE64_ENCODE
![Image](https://github.com/GoogleCloudPlatform/professional-services-data-validator/assets/37883833/36b83831-2386-499c-97f5-cfadf2887a09)

- HEX_ENCODE
![Image](https://github.com/GoogleCloudPlatform/professional-services-data-validator/assets/37883833/502981d1-5cd8-410d-b85f-e5179cf32805)

- Invalid argument types for function 'BASE64_DECODE_BINARY': (BINARY(16))
- Invalid argument types for function 'BASE64_DECODE_STRING': (BINARY(16))
- Invalid argument types for function 'HEX_DECODE_STRING': (BINARY(16))
- Invalid argument types for function 'HEX_DECODE_BINARY': (BINARY(16)) 

Snowflake reference: https://docs.snowflake.com/sql-reference/functions/base64_decode_binary

**Case 2**
It works now both with or without the BQ result handler and getting successful validations ![Image](https://github.com/GoogleCloudPlatform/professional-services-data-validator/assets/37883833/c8fc450b-8569-4ad8-84cd-543ecd3e246e)

YAML used
```yaml
result_handler:
  project_id: PROJECT_ID
  table_id: pso_data_validator.results
  type: BigQuery
source: snowflake
target: bq
validations:
- calculated_fields: []
  comparison_fields:
  - cast: null
    field_alias: dv_id
    source_column: DV_ID
    target_column: DV_ID # dv_id on issue
  filter_status: null
  filters: []
  format: table
  labels: []
  primary_keys:
  - cast: null
    field_alias: key_field
    source_column: KEY_FIELD
    target_column: KEY_FIELD # key_field on issue
  random_row_batch_size: '100'
  schema_name: PSO_DATA_VALIDATOR.PUBLIC
  table_name: ISSUE1070
  target_schema_name: pso_data_validator
  target_table_name: issue1070
  threshold: 0.0
  type: Row
  use_random_rows: true
```

**Case 3**
- Still getting `Error cannot use a string pattern on a bytes-like object` when cast is NULL
- With cast as STRING getting `Can not convert parameter 'X'64484B73565932494C47'' of type [BINARY(10)] into expected type [VARCHAR(32)]`

**Comment by helensilva14** [comment]
> Also if I run this one:
> 
> ```
> data-validation validate row \
>   -sc bq_conn \
>   -tc sf_conn \
>   -tbls dv.TABLE1=dv.TABLE1 \
>   --concat key_field \
>   --primary-keys dv_id \
>   -rr \
>   -rbs 100
> ```
> 
> It fails in: `TypeError: cannot use a string pattern on a bytes-like object`

Still getting this error as well in my current branch

**Comment by sundar-mudupalli-work** [comment]
Henri,

After further review, DVT is not handling binary data correctly - thank you for pointing this issue to us. There is a fairly straightforward workaround for this - use custom query validation. The DVT command to do the equivalent random row validation is
```
data-validation validate custom-query row -hash DV_ID -pk KEY_FIELD -rr -rbs 10 -sc snowflake -tc bq -sq 'select HEX_ENCODE(DV_ID) as DV_ID, KEY_FIELD from PSO_DATA_VALIDATOR.PUBLIC.ISSUE1070;' 
-tq 'select TO_HEX(DV_ID) as DV_ID, KEY_FIELD from pso_data_validator.issue1070;'
```
This will also work with BQ report handler.

Can you let us know if this addresses your issue(s) ?

Thanks.

Sundar Mudupalli

**Comment by henrihem** [comment]
Hi,
I'm well aware of the workaround to use custom query validations and that is what we've been using for simple cases.

However, it would be unfortunate not to be able to use query generation by DVT, as that would be really useful for our situation. Therefore I would still request DVT would support BINARY-keys.

I think currently DVT is trying to CAST the BINARY in Snowflake, but the casting is incorrect. I think this is a bug that needs to be fixed and the solution is to use HEX_ENCODE(dv_id, 0) in Snowflake.

Basically:
BigQuery:
`to_hex(dv_id)`
Snowflake:
`hex_encode(dv_id, 0)`
are equal and will create string-representation of the binary-key in lowercase.
Note that the casing in Snowflake needs to be set with 0.
